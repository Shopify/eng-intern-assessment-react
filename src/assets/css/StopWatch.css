.stopwatch-container {
    text-align: center;
    font-family: Verdana;
}

.title {
    padding-top: 8vh;
    font-size: 3em;
}

.time-display {
    padding: 50px 0;
    font-size: 5em;
    max-width: 100%;
}