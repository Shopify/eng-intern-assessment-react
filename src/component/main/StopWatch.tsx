import React, { useEffect, useRef, useState } from "react";
import StopWatchButton from "../common/button/StopWatchButton";
import Time from "../common/time/Time";
import Lap from "../common/lap/Lap";
import classes from "./StopWatch.module.css";
import btnclasses from "../common/button/StopWatchButton.module.css";

//Declare and initalize an initialTime for the stopwatch
const initialTime: TimeType = {
  milli: 0,
  second: 0,
  minute: 0,
  hour: 0,
};

//Declare and initalize an initialDateTime for the stopwatch
const initialDateTime: DateTimeType = {
  currentTime: Date.now(),
  pausedTime: Date.now(),
  capturedTime: Date.now(),
};

export default function StopWatch() {
  //Represents the initial state of the timer, true if first, false after
  const initialStartRef = useRef(true);

  //Represents the reset state. True on reset
  const onResetRef = useRef(false);
  //Holds the delay value of the laps. Incremented by 250 each lap
  const delayRef = useRef(0);
  //Represent the timer to reset lap compoenent. depend on delayRef value
  let lapsDelayTimer: NodeJS.Timeout;

  //State that holds the time of the stopwatch
  const [timeState, setTimeState] = useState({
    time: initialTime,
    lapTime: initialTime,
    dateTime: initialDateTime,
  });

  //State that holds the laps component array of the stopwatch
  const [laps, setLapsState] = useState<{
    lapNumber: number;
    component: TimeType[];
  }>({
    lapNumber: 0,
    component: [],
  });

  /**
   * State that holds the properties of each stopwatch button
   *
   * btn1: Represents Start button
   * btn2: Represents Stop button
   * btn3: Represents Lap button
   * btn4: Represents Reset button
   */
  const [btnState, setBtnState] = useState({
    className: "",
    btnType: {
      btn1: {
        label: "Start",
        isActive: true,
      },
      btn2: {
        label: "Stop",
        isActive: false,
      },
      btn3: {
        label: "Lap",
        isActive: true,
      },
      btn4: {
        label: "Reset",
        isActive: false,
      },
    },
  });

  //Function that handles button click
  const handleButtonClick = (btnClick: string) => {
    //Condition If Start button is click
    if (btnClick === "Start") {
      if (!btnState.btnType.btn2.isActive) {
        //Constantly check if stop button isn't active. It means it was paused
        //Calculate the pause Duration: d = (now - timeOnStop)
        const pauseDuration = Date.now() - timeState.dateTime.pausedTime;

        //Sets value
        initialStartRef.current = true;
        onResetRef.current = false;

        //Set the current and captured time to currentTime plus the pause time accumulated if it was pause
        setTimeState((prev) => {
          return {
            ...prev,
            dateTime: {
              ...prev.dateTime,
              currentTime: prev.dateTime.currentTime + pauseDuration,
              capturedTime: prev.dateTime.capturedTime + pauseDuration,
            },
          };
        });
      }
    }

    //Condition If Stop button is click
    //Set the pausedTime value to the time it was clicked
    if (btnClick === "Stop") {
      setTimeState((prev) => {
        return {
          ...prev,
          dateTime: {
            ...prev.dateTime,
            pausedTime: Date.now(),
          },
        };
      });
    }

    //Condition If Reset button is click
    //Set TimeState and Laps to intial value
    if (btnClick === "Reset") {
      if (btnState.btnType.btn4) {
        onResetRef.current = true;
        lapsDelayTimer = setTimeout(
          //Wait for the laps animation to finish before setting to initial value
          () => {
            setLapsState(() => {
              return {
                lapNumber: 0,
                component: [],
              };
            });
          },
          delayRef.current <= 1000 ? 3000 : delayRef.current + 1200
        );

        setTimeState(() => {
          return {
            time: initialTime,
            lapTime: initialTime,
            dateTime: initialDateTime,
          };
        });
      }
    }
    //Set the button state to know which button to activate or deactivate
    setBtnState((prev) => {
      switch (
        btnClick //Surround each with a switch case
      ) {
        case "Start":
          return {
            ...prev,
            btnType: {
              ...prev.btnType,
              btn1: {
                ...prev.btnType.btn1,
                isActive: false,
              },
              btn2: {
                ...prev.btnType.btn2,
                isActive: true,
              },
              btn3: {
                ...prev.btnType.btn3,
                isActive: true,
              },
            },
          };
        case "Stop":
          return {
            ...prev,
            btnType: {
              ...prev.btnType,
              btn1: {
                ...prev.btnType.btn1,
                isActive: true,
              },
              btn2: {
                ...prev.btnType.btn2,
                isActive: false,
              },
              btn3: {
                ...prev.btnType.btn3,
                isActive: false,
              },
            },
          };
        case "Lap":
          return {
            ...prev,
            btnType: {
              ...prev.btnType,
              btn3: {
                ...prev.btnType.btn3,
                isActive: true,
              },
            },
          };
        case "Reset":
          return {
            ...prev,
            btnType: {
              ...prev.btnType,
              btn3: {
                ...prev.btnType.btn3,
                isActive: true,
              },
              btn4: {
                ...prev.btnType.btn4,
                isActive: false,
              },
            },
          };
        default:
          return { ...prev };
      }
    });
  };

  //Clean up function
  useEffect(() => {
    return () => {
      clearTimeout(lapsDelayTimer);
    };
  }, []);

  //useEffect to track changes buttons and lap number then start the stopwatch timer
  useEffect(() => {
    const startTimer = setInterval(() => {
      //First timer to start the stopwatch timer
      if (btnState.btnType.btn2.isActive) {
        // If button start is click, button stop will be active
        setTimeState((prev) => {
          //Set the time state as follows

          //Caluclate the time difference from time it was started to Now (td = now - curr)
          const timeDifference = Date.now() - prev.dateTime.currentTime;

          //initialize a temp time
          //(dividing the time on millis by the number of millis)
          const temp: TimeType = {
            milli: prev.time.milli >= 1000 ? 0 : prev.time.milli + 10, //Increment 10, reset after 1000
            hour: Math.floor((timeDifference / (1000 * 60 * 60)) % 24), // Formula to calculate hour
            minute: Math.floor((timeDifference / (1000 * 60)) % 60), // Formula to calculate minute
            second: Math.floor((timeDifference / 1000) % 60), // Formula to calculate second
          };
          return {
            ...prev,
            time: temp,
          };
        });
      }
    }, 10);

    //Seocond timer representing the laps timer
    const startLapTimer = setInterval(() => {
      if (btnState.btnType.btn2.isActive) {
        setTimeState((prev) => {
          //Calculate time difference, if initial lap, get the difference from start time else from time lap button was click
          const timeDifference =
            laps.lapNumber === 0
              ? Date.now() - prev.dateTime.currentTime
              : Date.now() - prev.dateTime.capturedTime;
          const temp: TimeType = {
            milli: prev.time.milli >= 1000 ? 0 : prev.time.milli + 10,
            hour: Math.floor((timeDifference / (1000 * 60 * 60)) % 24),
            minute: Math.floor((timeDifference / (1000 * 60)) % 60),
            second: Math.floor((timeDifference / 1000) % 60),
          };

          return {
            ...prev,
            lapTime: temp,
          };
        });
      }
    }, 10);

    //Clear interval
    return () => {
      clearInterval(startLapTimer);
      clearInterval(startTimer);
    };
  }, [btnState.btnType.btn1, btnState.btnType.btn2, laps.lapNumber]);

  //useEffect to track lap button click
  useEffect(() => {
    if (initialStartRef.current) {
      initialStartRef.current = false;
      return;
    }

    //Run function if stop button is active
    if (btnState.btnType.btn2.isActive) {
      //Set lap state with lapNumber and laptime
      setLapsState((prev) => {
        return {
          lapNumber: prev.lapNumber + 1,
          component: [...prev.component, timeState.lapTime],
        };
      });

      //Reset the lap current time to start counting back to zero
      //Set the lap captured time to now
      setTimeState((prev) => {
        return {
          ...prev,
          lapTime: initialTime,
          dateTime: {
            ...prev.dateTime,
            capturedTime: Date.now(),
          },
        };
      });
    }
  }, [btnState.btnType.btn3]);

  //Hanlde which button to display for the first button
  const hanldeStartStop = btnState.btnType.btn1.isActive
    ? btnState.btnType.btn1
    : btnState.btnType.btn2;
  //Hanlde which button to display for the second button
  const hanldeLapReset = btnState.btnType.btn3.isActive
    ? btnState.btnType.btn3
    : btnState.btnType.btn4;

  //Component renders all the time, make sure delay ref is set to zero
  delayRef.current = 0;

  return (
    <div className="main">
      <div className="container">
        <div className={classes.time}>
          <Time time={timeState.time} className="stopwatch" />
        </div>
        <div title={"button"} className={btnclasses["btn-container"]}>
          <StopWatchButton
            btnType={hanldeStartStop}
            onButtonClick={handleButtonClick}
            className={hanldeStartStop.label}
            name={hanldeStartStop.label}
          />
          <StopWatchButton
            btnType={hanldeLapReset}
            onButtonClick={handleButtonClick}
            className={hanldeLapReset.label}
            name={hanldeLapReset.label}
          />
        </div>
      </div>
      <div title={"laps"} className={classes["lap-container"]}>
        <ul role={"ul"} className={classes.ul}>
          {laps.component.map((lap, i) => {
            return (
              <li role={"li"} key={i} className={classes.li}>
                <Lap
                  className={
                    onResetRef.current
                      ? i % 2 === 0
                        ? "startLeft"
                        : "startRight"
                      : "idle"
                  }
                  delay={(delayRef.current += 250)}
                  index={i}
                  lap={lap}
                />
              </li>
            );
          })}
        </ul>
      </div>
    </div>
  );
}
