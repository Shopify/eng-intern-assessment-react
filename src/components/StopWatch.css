body {
  background: rgb(218, 112, 214);
  background: radial-gradient(
    circle,
    rgba(218, 112, 214, 1) 0%,
    rgba(148, 187, 233, 1) 100%
  );
  margin: 0;
  display: flex;
  min-height: 100vh;
  justify-content: center;
  align-items: center;
}

.container {
  color: #f0a7ee;
  display: flex;
  letter-spacing: 2px;
  flex-wrap: wrap;
  text-align: center;
}

.container span {
  position: relative;
  font-size: 6rem;
  bottom: 3px;
}

.text {
  margin: 0;
  font-size: 6rem;
}

.btn {
  background-color: #d8bfd8;
  color: black;
  padding: 10px 20px;
  font-size: 1.5rem;
  margin: 2px 13px 10px;
  cursor: pointer;
  border: 1px solid purple;
  border-radius: 4px;
}

.btn:hover {
  background-color: #da70d6;
}

.btn.selected {
  background-color: #da70d6;
}

.lap-list {
  font-size: 2rem;
  text-align: center;
  margin-top: 2rem;
}

/* smaller screens */

@media only screen and (max-width: 600px) {
  .container span {
    font-size: 4rem;
  }

  .text {
    font-size: 4rem;
  }

  .btn {
    padding: 8px 16px;
    font-size: 1rem;
    margin: 5px;
  }
}
