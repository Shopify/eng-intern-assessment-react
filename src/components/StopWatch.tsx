import React, { useState, useEffect, useRef } from 'react';
import StopWatchButton from './StopWatchButton';

// StopWatch component - represents all the functions of a standard stopwatch
const StopWatch: React.FC = () => {
  const [isRunning, setIsRunning] = useState<boolean>(false);
  const [elapsedTime, setElapsedTime] = useState<number>(0);
  const [laps, setLaps] = useState<{ lapNumber: number; lapTime: number }[]>([]);
  
  // Used to store lap start times and lap number
  const lapStartTimeRef = useRef<number | null>(null);
  const lapNumberRef = useRef<number>(1);

  // Hook that updates the time when the stopwatch starts (update interval is 1 ms).
  useEffect(() => {
    let intervalId: NodeJS.Timeout;

    if (isRunning) {
      const startTime = Date.now() - elapsedTime;

      // Update elapsed time at a regular interval
      intervalId = setInterval(() => {
        const elapsed = Date.now() - startTime;
        setElapsedTime(elapsed);
      }, 1);
    }

    return () => clearInterval(intervalId);
  }, [isRunning, elapsedTime]);

  // Event that starts the stopwatch
  const handleStart = () => {
    setIsRunning(true);
    lapStartTimeRef.current = Date.now();
  };

  // Event that stops the stopwatch
  const handleStop = () => {
    setIsRunning(false);
  };

  // Event that resets the stopwatch
  const handleReset = () => {
    setIsRunning(false);
    setElapsedTime(0);
    setLaps([]);
    lapNumberRef.current = 1;
    lapStartTimeRef.current = null;
  };

  // Event that records a lap only when the stopwatch is running. It records the difference from the last saved start time.
  const handleLap = () => {
    if (isRunning && lapStartTimeRef.current !== null) {
      const lapTime = Date.now() - lapStartTimeRef.current;
      const lapNumber = lapNumberRef.current;

      lapNumberRef.current = lapNumber + 1;

      setLaps((prevLaps) => [
        { lapNumber, lapTime },
        ...prevLaps.slice(0, 9), 
      ]);

      lapStartTimeRef.current = Date.now();
    }
  };

  // Format the time in as 00:00:00.000 (hh:mm:ss.ms)
  const formatTime = (time: number) => {
    const date = new Date(time);
    const hours = date.getUTCHours().toString().padStart(2, '0');
    const minutes = date.getUTCMinutes().toString().padStart(2, '0');
    const seconds = date.getUTCSeconds().toString().padStart(2, '0');
    const milliseconds = date.getUTCMilliseconds().toString().padStart(3, '0');
    return `${hours}:${minutes}:${seconds}.${milliseconds}`;
  };

  return (
    <div>
      <h1>{formatTime(elapsedTime)}</h1>
      <StopWatchButton onClick={handleStart} label="Start" />
      <StopWatchButton onClick={handleStop} label="Stop" />
      <StopWatchButton onClick={handleReset} label="Reset" />
      <StopWatchButton onClick={handleLap} label="Lap" />

      {laps.length > 0 && (
        <div>
          <h2>Lap Times</h2>
          <ul>
            {laps.map(({ lapNumber, lapTime }, index) => (
              <li key={index}>{`Lap ${lapNumber}: ${formatTime(lapTime)}`}</li>
            ))}
          </ul>
        </div>
      )}
    </div>
  );
};

export default StopWatch;