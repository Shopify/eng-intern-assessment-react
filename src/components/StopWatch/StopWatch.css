.top {
  display: flex;
  flex-direction: column;
  justify-content: flex-end;
  height: 50%;
  align-items: center;
  padding-bottom: 25px;
  border-bottom: 1px solid #242427;
}

.stopwatch-text {
  color: white;

  display: flex;
  flex-direction: row;
  font-size: 52px;
  letter-spacing: 2px;
}

.bottom {
  display: flex;
  align-items: flex-start;
  height: 50%;
  overflow: scroll;
  justify-content: center;
}
