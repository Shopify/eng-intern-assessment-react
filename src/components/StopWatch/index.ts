export * from "./hooks/useStopWatch"
export * from "./DigitalDisplay/"
export * from "./AnalogDisplay/"
export * from "./StopWatchWrapper"
export * from "./utils"
export * from "./StopWatchButton/"
