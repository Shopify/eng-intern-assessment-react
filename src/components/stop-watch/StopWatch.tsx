import React, { useEffect } from 'react';
import { useState } from 'react';
import { StopWatchProps } from '../../props/StopWatchProps';
import StopWatchButton from './StopWatchButton';
import { StopWatchLap } from './StopWatchLap';
import { containerStyles, titleStyle } from '../../styles/StopWatchStyles';

export default function StopWatch(
    {
        id = Math.random(),
        title,
        time = 0,
        timelaps = []
    }: StopWatchProps) {

    // Hooks:
    const [runningTime, setRunningTime] = useState(Math.max(time, 0));
    const [displayTime, setDisplayTime] = useState("00:00:00");
    const [increment, setIncrement] = useState(10); // increment of increase in milliseconds
    const [isRunning, setIsRunning] = useState(false);
    const [laps, setLaps] = useState(timelaps);

    // Constants:
    const MAX_LAPS_DISPLAY_NUM = 5; // maximum laps displayed

    // Functions:
    const IncrementTimer = () => {
        setRunningTime(priorTime => priorTime + increment);
    };

    const StartTimer = () => {
        setIsRunning(true);
    };

    const StopTimer = () => {
        setIsRunning(false);
    };

    const ResetTimer = () => {
        StopTimer();
        setRunningTime(0);
    };

    const AddLap = () => {
        setLaps(prevLaps => [...prevLaps, { displayTime: displayTime, time: runningTime }]);
    };

    const UpdateDisplayTime = () => {
        const miliseconds = Math.floor((runningTime / 10) % 100);
        const seconds = Math.floor((runningTime / 1000) % 60);
        const minutes = Math.floor(seconds / 60);

        const display = (minutes / 10 < 1 ? "0" : "") + String(minutes) + ":"
            + (seconds / 10 < 1 ? "0" : "") + String(seconds) + ":" + (miliseconds / 10 < 1 ? "0" : "") + String(miliseconds);

        setDisplayTime(display);
    };

    // Set displayTime on component first render
    useEffect(() => {
        UpdateDisplayTime();
    }, []);

    // Update displayTime accordingly to runningTime
    useEffect(() => {
        UpdateDisplayTime();
    }, [runningTime]);

    /**
     * Update timer if isRunning is true,
     * else stop the timer.
     * */
    useEffect(() => {
        let interval: number | string | NodeJS.Timeout;

        if (isRunning) {
            interval = setInterval(() => {
                IncrementTimer();
            }, 10);
        } else {
            clearInterval(interval);
        };

        return () => { clearInterval(interval) };

    }, [isRunning]);


    return (
        <div>
            <div className='stopwatch-title' style={titleStyle}>{title}</div>
        <div className='stopwatch-container' style={containerStyles.stopwatchContainer}>
            <div className='stopwatch-timer' style={containerStyles.stopwatchTimer} title='timer'>
                {displayTime}
            </div>
            <div className='stopwatch-button-container' style={containerStyles.stopwatchButtonContainer}>
                <StopWatchButton title='Start' callback={StartTimer} />
                <StopWatchButton title='Stop' callback={StopTimer} />
                <StopWatchButton title='Reset' callback={ResetTimer} />
                <StopWatchButton title='Lap' callback={AddLap} />
            </div>
            <div className='stopwatch-lap-container' style={containerStyles.stopwatchLapContainer} title='laps'>
                {
                    // display most recent MAX_LAPS_DISPLAY_NUM laps
                    laps.slice(0).reverse().slice(0, MAX_LAPS_DISPLAY_NUM).map(
                        (lap) => { return <StopWatchLap key={lap.time} time={lap.time} displayTime={lap.displayTime} /> })
                }
            </div>
        </div>
        </div>
    );
};
