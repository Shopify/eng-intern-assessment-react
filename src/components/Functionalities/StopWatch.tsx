import React from 'react'
import './StopWatch.scss'
import StopWatchButton from '../Button/StopWatchButton'

export default function StopWatch() {
    return(
        <div className="watch__glass"></div>
    )
}