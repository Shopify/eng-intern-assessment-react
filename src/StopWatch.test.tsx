import React from 'react';
import { render, screen, fireEvent, act } from '@testing-library/react';
import StopWatch from './StopWatch';
import '@testing-library/jest-dom';

// Setup: Mocking timers for the setInterval used in the StopWatch component
beforeEach(() => {
    jest.useFakeTimers();
});

afterEach(() => {
    jest.useRealTimers();
});

// Test suite for the StopWatch Component
describe('StopWatch Component', () => {
    // Test case for start/stop functionality
    test('should start and stop counting when start/stop button is clicked', () => {
        render(<StopWatch />);
        const startStopButton = screen.getByText('Start/Stop');

        // Start the stopwatch and advance time by 1 second
        fireEvent.click(startStopButton);
        act(() => {
            jest.advanceTimersByTime(1000);
        });
        let display = screen.getByText(/0:00:01/);
        expect(display).toBeInTheDocument();

        // Stop the stopwatch and ensure time remains the same
        fireEvent.click(startStopButton);
        act(() => {
            jest.advanceTimersByTime(1000);
        });
        display = screen.getByText(/0:00:01/);
        expect(display).toBeInTheDocument();
    });

    // Test case for lap recording functionality
    test('should record lap times correctly', () => {
        render(<StopWatch />);
        const startStopButton = screen.getByText('Start/Stop');
        const lapButton = screen.getByText('Lap');

        // Start the stopwatch, advance time, and record a lap
        fireEvent.click(startStopButton);
        act(() => {
            jest.advanceTimersByTime(5000);
        });
        fireEvent.click(lapButton);

        // Verify that lap times are displayed correctly
        const lapTimes = screen.getAllByText(/0:00:05/);
        expect(lapTimes.length).toBe(2);
    });

    // Test case for reset functionality
    test('should reset the time and laps when reset button is clicked', () => {
        render(<StopWatch />);
        const startStopButton = screen.getByText('Start/Stop');
        const resetButton = screen.getByText('Reset');

        // Start the stopwatch, advance time, and then reset
        fireEvent.click(startStopButton);
        act(() => {
            jest.advanceTimersByTime(1000);
        });
        fireEvent.click(resetButton);

        // Verify that the stopwatch resets to zero
        const display = screen.getByText(/0:00:00/);
        expect(display).toBeInTheDocument();
    });

    // Test case for time display format
    test('should display time in the correct format', () => {
        render(<StopWatch />);
        const startStopButton = screen.getByText('Start/Stop');

        // Start the stopwatch and advance time to check format
        fireEvent.click(startStopButton);
        act(() => {
            jest.advanceTimersByTime(3661000); // 1 hour, 1 minute, and 1 second
        });

        // Verify that time is displayed in the correct format
        const display = screen.getByText(/01:01:01/);
        expect(display).toBeInTheDocument();
    });
});