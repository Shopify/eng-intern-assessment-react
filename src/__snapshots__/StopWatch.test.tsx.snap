// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`StopWatch tests render should match snapshot file 1`] = `
<div
  className="clock"
>
  <div>
    00:00:00.00
  </div>
  <div
    className="buttons"
  >
    <button
      className="btn"
      disabled={false}
      onClick={[Function]}
    >
      Start
    </button>
    <button
      className="btn"
      disabled={true}
      onClick={[Function]}
    >
      Stop
    </button>
                    
    <button
      className="btn"
      disabled={true}
      onClick={[Function]}
    >
      Lap
    </button>
    <button
      className="btn"
      disabled={false}
      onClick={[Function]}
    >
      Reset
    </button>
  </div>
  <div
    className="lap"
  />
</div>
`;
